controlplane/kubeadm/internal/machinefilters/machine_filters.go

```diff
@@ -197,3 +202,87 @@ func ControlPlaneSelectorForCluster(clusterName string) labels.Selector {
+func MatchesKCPConfiguration(ctx context.Context, c client.Client, kcp controlplanev1.KubeadmControlPlane) func(machine *clusterv1.Machine) bool {
+	return And(
+		MatchesKubernetesVersion(kcp.Spec.Version),
+		MatchesKubeadmBootstrapConfig(ctx, c, kcp.Spec.KubeadmConfigSpec),
+		MatchesTemplateClonedFrom(ctx, c, kcp),
+	)
+}
+
+// MatchesTemplateClonedFrom returns a filter to find all machines that match a given KCP infra template.
+func MatchesTemplateClonedFrom(ctx context.Context, c client.Client, kcp controlplanev1.KubeadmControlPlane) Func {
+	return func(machine *clusterv1.Machine) bool {
+		if machine == nil {
+			return false
+		}
+		infraObj, err := external.Get(ctx, c, &machine.Spec.InfrastructureRef, kcp.Namespace)
+		// Return true here because failing to get infrastructure machine should not be considered as unmatching.
+		if err != nil {
+			return true
+		}
+
+		clonedFromName, ok1 := infraObj.GetAnnotations()[clusterv1.TemplateClonedFromNameAnnotation]
+		clonedFromGroupKind, ok2 := infraObj.GetAnnotations()[clusterv1.TemplateClonedFromGroupKindAnnotation]
+		if !ok1 || !ok2 {
+			// All kcp cloned infra machines should have this annotation.
+			// Missing the annotation may be due to older version machines or adopted machines.
+			// Should not be considered as mismatch.
+			return true
+		}
+
+		//Check if infra machine's InfrastructureTemplate is matching the kcp InfrastructureTemplate.
+		if clonedFromName != kcp.Spec.InfrastructureTemplate.Name ||
+			clonedFromGroupKind != kcp.Spec.InfrastructureTemplate.GroupVersionKind().GroupKind().String() {
+			return false
+		}
+		return true
+	}
+}
+
+// MatchesKubernetesVersion returns a filter to find all machines that match a given Kubernetes version.
+func MatchesKubernetesVersion(kubernetesVersion string) Func {
+	return func(machine *clusterv1.Machine) bool {
+		if machine == nil {
+			return false
+		}
+		return *machine.Spec.Version == kubernetesVersion
+	}
+}
+
+// MatchesKubeadmBootstrapConfig checks if machine's KubeadmConfigSpec is equivalent with KCP's KubeadmConfigSpec.
+func MatchesKubeadmBootstrapConfig(ctx context.Context, c client.Client, kcpConfigSpec bootstrapv1.KubeadmConfigSpec) Func {
+	return func(machine *clusterv1.Machine) bool {
+		bootstrapRef := machine.Spec.Bootstrap.ConfigRef
+		if bootstrapRef == nil {
+			// Missing bootstrap reference should not be considered as unmatching.
+			return true
+		}
+
+		bootstrapObj := &bootstrapv1.KubeadmConfig{}
+		if err := c.Get(ctx, client.ObjectKey{Name: bootstrapRef.Name, Namespace: machine.Namespace}, bootstrapObj); err != nil {
+			// Return true here because failing to get KubeadmConfig should not be considered as unmatching.
+			return true
+		}
+
+		// TODO: do we need this check?
+		if bootstrapObj.Spec.InitConfiguration == nil && bootstrapObj.Spec.JoinConfiguration == nil && bootstrapObj.Spec.ClusterConfiguration == nil {
+			return false
+		}
```

Comment:
Good question, not sure exactly how we should treat this case off the top of my head, as it could mean that it was created with the expectation that a default InitConfiguration would be created, but would not have been reconciled yet to have the default InitConfiguration available yet.

Reply:
I can't think of any benefit of forcing to upgrade machines that has all 3 configs missing, because as you said in the next reconcile they will have some defaulted values, so removing this check.